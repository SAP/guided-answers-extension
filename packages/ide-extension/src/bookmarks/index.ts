export { getAllBookmarks, initBookmarks, updateBookmarks } from './bookmarks';
