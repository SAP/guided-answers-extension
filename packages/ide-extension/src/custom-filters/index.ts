export { getAllCustomFilters, initCustomFilters, updateCustomFilters } from './customFilters';
