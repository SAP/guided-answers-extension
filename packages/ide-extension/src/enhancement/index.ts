export { handleCommand } from './commandHandler';
export { getEnhancements } from './enhancements';
