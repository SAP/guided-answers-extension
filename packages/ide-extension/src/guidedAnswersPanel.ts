import { Uri, ViewColumn, WebviewPanel, window, workspace } from 'vscode';
// import { ChannelName, ExtensionLogger } from '@sap/ux-extension-logger';
import {
    GuidedAnswerActions,
    GuidedAnswerAPI,
    SELECT_NODE,
    updateGuidedAnserTrees,
    updateActiveNode,
    EXECUTE_COMMAND,
    SEARCH_TREE,
    WEBVIEW_READY,
    setActiveTree,
    HTMLEnhancement,
    NodeEnhancement
} from '@sap/guided-answers-extension-types';
import { getGuidedAnswerApi } from '@sap/guided-answers-extension-core';
// import { webappDirPath } from '@sap/ux-guided-answer-webapp';
import { getHtml } from './utils/web';
import { uriToResourceUri } from './utils/vscode-util';
import { handleCommand } from './commandHandler';

// const logger = new ExtensionLogger(ChannelName.Help);

export class GuidedAnswersPanel {
    // public readonly panel: WebviewPanel;
    private guidedAnswerApi: GuidedAnswerAPI;
    private readonly initialTree: number | undefined;

    constructor(treeId?: number) {
        try {
            this.initialTree = treeId;
            const config = workspace.getConfiguration('sap.ux.guidedAnswer');
            const apiHost = config.get('apiHost') as string;
            const nodeEnhancements = config.get('nodeEnhancements') as NodeEnhancement[];
            const htmlEnhancements = config.get('htmlEnhancements') as HTMLEnhancement[];
            const enhancements = {
                nodeEnhancements,
                htmlEnhancements
            };
            this.guidedAnswerApi = getGuidedAnswerApi({ apiHost, enhancements });
            // const webAppUri = Uri.file(webappDirPath);
            // this.panel = window.createWebviewPanel('sap.ux.guidedAnswer.view', 'Guided Answers', ViewColumn.Active, {
            //     enableCommandUris: true,
            //     enableScripts: true,
            //     retainContextWhenHidden: true,
            //     localResourceRoots: [Uri.file(webappDirPath)],
            //     enableFindWidget: true
            // });
            // this.panel.title = 'Guided Answers';
            // this.panel.webview.onDidReceiveMessage(this.onWebviewMessage.bind(this));
            // const html = getHtml(
            //     uriToResourceUri(webAppUri),
            //     'Guided Answers',
            //     '/guidedAnswers.js',
            //     undefined,
            //     '/guidedAnswers.css'
            // );
            // this.panel.webview.html = html;
        } catch (error) {
            // logger.error(error.message);
            throw error;
        }
    }

    /**
     * Handler for actions coming from webview. This should be primaraly commands with arguments
     * @param action - action to execute
     */
    private onWebviewMessage(action: GuidedAnswerActions): void {
        try {
            switch (action.type) {
                case SELECT_NODE: {
                    this.guidedAnswerApi.getNodeById(action.payload).then((node) => {
                        this.postActionToWebview(updateActiveNode(node));
                    });
                    break;
                }
                case EXECUTE_COMMAND: {
                    handleCommand(action.payload);
                    break;
                }
                case SEARCH_TREE: {
                    this.guidedAnswerApi
                        .getTrees(action.payload)
                        .then((trees) => this.postActionToWebview(updateGuidedAnserTrees(trees)));
                    break;
                }
                case WEBVIEW_READY: {
                    if (this.initialTree) {
                        this.guidedAnswerApi.getTreeById(this.initialTree).then((tree) => {
                            this.guidedAnswerApi.getNodeById(tree.FIRST_NODE_ID).then((node) => {
                                this.postActionToWebview(setActiveTree(tree));
                                this.postActionToWebview(updateActiveNode(node));
                            });
                        });
                    }
                    break;
                }
            }
        } catch (error) {
            // logger.error(`Error processing message`, action, error);
        }
    }

    /**
     * This should be the ONLY function that sends data to application info webview
     * @param action - the action to post to application info webview
     */
    private postActionToWebview(action: GuidedAnswerActions): void {
        // this.panel.webview.postMessage(action);
    }

    /**
     * Show Application Info panel and set application info data
     */
    public show(): void {
        // this.panel.reveal();
    }
}
