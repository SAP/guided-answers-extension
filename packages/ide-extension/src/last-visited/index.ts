export { getAllLastVisitedGuides, initLastVisited, updateLastVisitedGuides } from './lastVisited';
