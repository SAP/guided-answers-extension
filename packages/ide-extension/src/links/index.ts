export { GuidedAnswersUriHandler } from './guided-answers-uri-handler';
export { extractLinkInfo, generateExtensionLink, generateWebLink } from './link-info';
