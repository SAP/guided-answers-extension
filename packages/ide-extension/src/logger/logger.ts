import type { OutputChannel } from 'vscode';
import { window } from 'vscode';

let channel: OutputChannel;

/**
 * Log a message to the output console.
 *
 * @param message - log message
 */
export function logString(message: string): void {
    if (!channel) {
        channel = window.createOutputChannel(`Guided Answers Extension`);
    }
    channel.appendLine(message);
}
