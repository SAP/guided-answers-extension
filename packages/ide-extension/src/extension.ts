import type { ExtensionContext } from 'vscode';
import { commands, window } from 'vscode';
import { getIde } from '@sap/guided-answers-extension-core';
import { logString } from './logger/logger';
import { GuidedAnswersPanel } from './panel/guidedAnswersPanel';
import type { StartOptions } from './types';

/**
 *  Activate function is called by VSCode when the extension gets active.
 *
 * @param context - context from VSCode
 */
export function activate(context: ExtensionContext): void {
    context.subscriptions.push(
        commands.registerCommand('sap.ux.guidedAnswer.openGuidedAnswer', async (startOptions?: StartOptions) => {
            try {
                const options = {
                    startOptions,
                    ide: await getIde()
                };
                logString(`Guided Answers command called. Options: ${JSON.stringify(options)}`);
                const guidedAnswersPanel = new GuidedAnswersPanel(options);
                guidedAnswersPanel.show();
            } catch (error) {
                window.showErrorMessage(`Error while starting Guided Answers: ${(error as Error).message}`);
            }
        })
    );
}
