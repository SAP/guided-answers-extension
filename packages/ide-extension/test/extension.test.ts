import { join } from 'path';
import { URI } from 'vscode-uri';
import { ExtensionContext, commands, window, WebviewPanel } from 'vscode';
import * as logger from '../src/logger/logger';
import { activate } from '../src/extension';

describe('Extension test', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('activate is function', () => {
        expect(typeof activate === 'function').toBeTruthy();
    });

    test('activate extension', () => {
        // Mock setup
        const subscriptionsMock = jest.spyOn(commands, 'registerCommand');
        const context = {
            subscriptions: []
        };

        // Test execution
        activate(context as unknown as ExtensionContext);

        // Result check
        expect(subscriptionsMock.mock.calls[0][0]).toBe('sap.ux.guidedAnswer.openGuidedAnswer');
        expect(typeof context.subscriptions[0]).toBe('function');
    });

    test('execute command', async () => {
        // Mock setup
        const loggerMock = jest.spyOn(logger, 'logString').mockImplementation(() => null);
        const subscriptionsMock = jest.spyOn(commands, 'registerCommand');
        const webViewPanelMock = {
            webview: {
                html: '',
                onDidReceiveMessage: jest.fn(),
                asWebviewUri: jest.fn().mockReturnValue(''),
                cspSource: ''
            },
            onDidChangeViewState: jest.fn(),
            onDidDispose: jest.fn(),
            reveal: jest.fn()
        } as unknown as WebviewPanel;
        jest.spyOn(window, 'createWebviewPanel').mockImplementation(() => webViewPanelMock);
        const context = {
            subscriptions: []
        };

        // Test execution
        activate(context as unknown as ExtensionContext);
        await subscriptionsMock.mock.calls[0][1]();

        // Result check
        const replaceStr = URI.file(join(__dirname, '..')).with({ scheme: 'vscode-resource' }).toString();
        expect(
            webViewPanelMock.webview.html.replace(
                new RegExp(`${replaceStr}`.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g'),
                ''
            )
        ).toMatchSnapshot();
        expect(webViewPanelMock.reveal).toBeCalled();
        expect(loggerMock).toBeCalled();
    });

    test('execute command with parameters', async () => {
        // Mock setup
        const loggerMock = jest.spyOn(logger, 'logString').mockImplementation(() => null);
        const subscriptionsMock = jest.spyOn(commands, 'registerCommand');
        const context = {
            subscriptions: []
        };

        // Test execution
        activate(context as unknown as ExtensionContext);
        await subscriptionsMock.mock.calls[0][1]({ treeId: 0, nodeIdPath: [1, 2, 3] });

        // Result check
        expect(loggerMock.mock.calls[0][0]).toContain('{"treeId":0,"nodeIdPath":[1,2,3]}');
    });

    test('execute command error occurs', async () => {
        // Mock setup
        jest.spyOn(logger, 'logString').mockImplementation(() => {
            throw Error('ERROR');
        });
        const showErrorMessageMock = jest.spyOn(window, 'showErrorMessage');
        const subscriptionsMock = jest.spyOn(commands, 'registerCommand');
        const context = {
            subscriptions: []
        };

        // Test execution
        activate(context as unknown as ExtensionContext);
        await subscriptionsMock.mock.calls[0][1]();

        // Result check
        expect(showErrorMessageMock.mock.calls[0][0]).toContain('ERROR');
    });
});
