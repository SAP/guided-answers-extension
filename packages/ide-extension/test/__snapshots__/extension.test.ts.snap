// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension test execute command 1`] = `
"<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="google" content="notranslate">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>Guided Answers</title>

        <base href="./">
        <link rel="stylesheet" type="text/css" href="/guidedAnswers.css">
        <script type="text/javascript">
            function onResourceLoadError() {
                // Load failed - send command to restart webview
                try {
                    var vscodeMessenger = window.vscode || acquireVsCodeApi();
                    if (vscodeMessenger) {
                        vscodeMessenger.postMessage({"type": "RESTART_WEBVIEW"});
                    }
                    // Rewrite function to avoid multiple restarts - restart logic should be callod only once
                    onResourceLoadError = function(){};
                } catch(e) {
                    console.log(e);
                }
            }
        </script>
        
    </head>
    <body>
        <noscript>You need to enable JavaScript to run this app.</noscript>
        <div id="root"></div>
        <script src="/guidedAnswers.js" onerror="onResourceLoadError()"></script>
    </body>
</html>
"
`;
